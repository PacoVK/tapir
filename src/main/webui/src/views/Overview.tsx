import React, {
  useCallback,
  useEffect,
  useLayoutEffect,
  useRef,
  useState,
} from "react";
import {
  Box,
  CircularProgress,
  List,
  Paper,
  Stack,
  TextField,
  Typography,
} from "@mui/material";
import ModuleElement from "../components/list/ModuleElement";
import { Module } from "../types";
import useDebounce from "../hooks/useDebounce";
import tapirLogo from "../assets/tapir.png";

const fetchDataLimit = 5;
const Overview = () => {
  const modulesTable = useRef();
  const [modules, setModules] = useState([] as Module[]);
  const [lastEvaluatedItemKey, setLastEvaluatedItemKey] = useState("");
  const [searchString, setSearchString] = useState("");
  const debouncedSearchTerm = useDebounce(searchString, 500);

  const [loading, setLoading] = useState(false);
  const [distanceBottom, setDistanceBottom] = useState(0);

  const hasMoreData = (lastEvaluatedItem: any) => {
    return (
      !!lastEvaluatedItem &&
      lastEvaluatedItem !== "" &&
      modules.at(0)?.id !== lastEvaluatedItem.id
    );
  };

  const loadMore = useCallback(
    () => {
      setLoading(true);
      fetchModules(
        `search?limit=${fetchDataLimit}&lastKey=${lastEvaluatedItemKey}&q=${searchString}`
      ).then((data) => {
        const allModules = [...modules, ...data.modules];
        setLastEvaluatedItemKey(
          data.lastEvaluatedItem ? data.lastEvaluatedItem.id : ""
        );
        setModules(allModules);
        setLoading(false);
      });
    },
    // eslint-disable-next-line
    [modules]
  );

  const scrollListener = useCallback(
    () => {
      let bottom =
        // @ts-ignore
        modulesTable.current.scrollHeight - modulesTable.current.clientHeight;

      if (!distanceBottom) {
        setDistanceBottom(Math.round(bottom * 0.2));
      }

      if (
        // @ts-ignore
        modulesTable.current.scrollTop > bottom - distanceBottom &&
        !loading &&
        lastEvaluatedItemKey !== ""
      ) {
        loadMore();
      }
    },
    // eslint-disable-next-line
    [loadMore, loading, distanceBottom]
  );

  useLayoutEffect(() => {
    const tableRef = modulesTable.current;
    // @ts-ignore
    tableRef.addEventListener("scroll", scrollListener);
    return () => {
      // @ts-ignore
      tableRef.removeEventListener("scroll", scrollListener);
    };
  }, [scrollListener]);

  useEffect(
    () => {
      setLoading(true);
      fetchModules(`search?limit=${fetchDataLimit}&q=${searchString}`).then(
        (data) => {
          setLastEvaluatedItemKey(
            data.lastEvaluatedItem ? data.lastEvaluatedItem.id : ""
          );
          setModules(data.modules);
          setLoading(false);
        }
      );
    },
    // eslint-disable-next-line
    [debouncedSearchTerm]
  );

  const fetchModules = async (api: string) => {
    const response = await fetch(api);
    return await response.json();
  };

  const handleSearchInputChange = (event: {
    target: { value: React.SetStateAction<string> };
  }) => {
    setSearchString(event.target.value);
  };

  const renderLastItem = () => {
    if (!modules || modules.length === 0) {
      return null;
    }
    return hasMoreData(lastEvaluatedItemKey) ? null : (
      <Paper>
        <Typography textAlign={"center"}>No more modules found</Typography>
      </Paper>
    );
  };

  return (
    <>
      <TextField
        id="module-search"
        label="Search module"
        type="search"
        value={searchString}
        onChange={handleSearchInputChange}
      />
      <List
        component={Stack}
        sx={{ maxHeight: "50vh", overflow: "auto", display: "flex-grow" }}
        // @ts-ignore
        ref={modulesTable}
      >
        {modules && modules.length > 0 ? (
          modules.map((module) => (
            <ModuleElement
              key={`${module.namespace}${module.name}${module.provider}`}
              module={module}
            />
          ))
        ) : !loading ? (
          <Box sx={{ margin: "auto" }}>
            <img alt={"Tapir logo"} src={tapirLogo} />
            <Typography textAlign={"center"}>
              Found this Tapir, but no modules
            </Typography>
          </Box>
        ) : null}
        {renderLastItem()}
        {loading ? (
          <CircularProgress color={"secondary"} sx={{ margin: "auto" }} />
        ) : null}
      </List>
    </>
  );
};

export default Overview;
