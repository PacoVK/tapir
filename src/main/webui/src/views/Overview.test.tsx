import React from "react";
import { render } from "@testing-library/react";
import Overview from "./Overview";

describe("<Overview /> spec", () => {
  it("renders the overview", () => {
    const container = render(<Overview />);
    expect(container).toMatchSnapshot();
  });
});
