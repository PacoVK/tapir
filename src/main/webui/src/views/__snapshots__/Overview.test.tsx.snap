// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Overview /> spec renders the overview 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="module-search"
          id="module-search-label"
        >
          Search module
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputTypeSearch css-1g24dm6-MuiInputBase-input-MuiOutlinedInput-input"
            id="module-search"
            type="search"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Search module
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiList-root MuiList-padding css-1475wex-MuiStack-root-MuiList-root"
      >
        <span
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorSecondary css-y6srfl-MuiCircularProgress-root"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="module-search"
        id="module-search-label"
      >
        Search module
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputTypeSearch css-1g24dm6-MuiInputBase-input-MuiOutlinedInput-input"
          id="module-search"
          type="search"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-yjsfm1"
          >
            <span>
              Search module
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiList-root MuiList-padding css-1475wex-MuiStack-root-MuiList-root"
    >
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorSecondary css-y6srfl-MuiCircularProgress-root"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
